### PR TITLE
CPP-212 - WIP expose SASL authentication (Do not merge!)

### PR DESCRIPTION
**Pull request by mpenick**

### CODE REVIEW

File: docs/topics/security/authentication/README.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                       size_t response_size) {
+  /*
+   * This callback is used to initiate a request to begin an authentication
+   * exchange. Required resources can be acquired and initialized here.
+   *
+   * Resources required for this specific exchange can be stored in the
+   * `auth->data` field and will be available in the subsequent challenge
+   * and success phases of the exchange. The cleanup callback should be used to
+   * free these resources. This data will only ever be accessed by a single
+   * thread.
```
Comment (aholmberg) [inline docs/topics/security/authentication/README.md]: single thread access: this seems to be contradicted by the Multithread comment in ``main()`` below. https://github.com/datastax/cpp-driver/pull/144/files#diff-8fa0bb629c64393413a846a276378027R102

File: include/cassandra.h [hunk 2]
```diff
@@ -483,11 +483,96 @@ typedef struct CassLogMessage_ {
...
+/**
+ * A callback used to cleanup resources that were acquired during
+ * the process of the authentication exchange. This is called after
+ * the termination of the exchange regardless of the outcome.
+ *
+ * @param[in] auth
+ * @param[in] data
+ */
+typedef void (*CassAuthCleanupCallback)(CassAuth* auth,
+                                        void* data);
```
Comment (aholmberg) [inline include/cassandra.h]: Do we want a comment about data threaded access here, in the api?

File: docs/topics/security/authentication/README.md (hunk 1 above)
Comment (mpenick) [inline docs/topics/security/authentication/README.md]: Hmmm. The  the `data` argument is different from the `auth->data`. I need to figure out how to make that not confusing....

File: docs/topics/security/authentication/README.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                         size_t response_size) {
+  /*
+   * This is used for handling an authentication challenge initiated
+   * by the server. The information contained in the token parameter is
+   * authentication protocol specific. It may be NULL or empty.
+   */
+
+   /*
+    * Return the number of bytes copied into `response` or the required size
+    * of `response` if it's too small.
```
Comment (aholmberg) [inline docs/topics/security/authentication/README.md]: Worth mentioning that the function will be called again if required size is > response_size? 
I think this is explained well in the API header.

File: docs/topics/security/authentication/README.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  return num_bytes;
+}
+
+void on_auth_success(CassAuth* auth,
+                     void* data,
+                     const char* token,
+                     size_t token_size ) {
+  /*
+   * This is to be used for handling the success phase of an exchange. The
+   * token parameters contains information that may be used to finialize
```
Comment (aholmberg) [inline docs/topics/security/authentication/README.md]: token *parameter* contains (singular)

File: src/auth.cpp [hunk 5]
```diff
@@ -42,4 +47,65 @@ void PlainTextAuthenticator::on_authenticate_success(const std::string& token) {
...
+
+std::string SaslAuthenticator::initial_response() {
+  if (callbacks_->initial_callback == NULL) {
+    return std::string();
+  }
+
+  std::string response(SASL_AUTH_INIT_RESPONSE_SIZE, '\0');
+  for (int i = 0; i < 2; ++i) {
+    size_t size = callbacks_->initial_callback(&auth_, data_,
+                                               &response[0], response.size());
```
Comment (aholmberg) [inline src/auth.cpp]: Just curious why ``&response[0]`` for string instead of ``response.data()`` in these calls.

**Comment by aholmberg** [comment]
This looks really good. 
Just put a few minor nit pick comments in the diff.

File: docs/topics/security/authentication/README.md (hunk 1 above)
Comment (aholmberg) [inline docs/topics/security/authentication/README.md]: You reminded me of another comment I forgot to mention: Is CassAuth.data ever used for anything, or is it always SaslAuthenticator.data_?

**Comment by yippeeH** [comment]
Hi Michael, it's great to see you guys have this working in progress. I'm just wondering what the status of this right now? Do you guys have any plan when to push it out? I need this and am really waiting on it. Appreciate it!  - Yinping

**Comment by mpenick** [comment]
@yippeeH This feature is code complete (minus a couple code review changes) and it needs to be formally tested. We haven't released this feature because there wasn't any demand for it and we didn't have an integrator to give us feedback about the API.

What type of authentication system are you planning to integrate with the SASL API? Kerberos? If I merge this on to the latest would you be able to test your integration from an unofficial release and provide feedback?

**Comment by yippeeH** [comment]
We are using a custom authentication implementation that is not a defined SASL mechanism, but requires a custom initial response callback from the driver. We would definitely be willing to test from an unofficial release and provide feedback.

**Comment by mpenick** [comment]
I'm moving this work to another branch and will reopen a new PR. I will reference this.

**Pull request closed without merging**